rama-http/src/service/web/router.rs

```diff
@@ -0,0 +1,512 @@
+        self.match_route(path, matcher, service)
+    }
+
+    /// add a POST route to the router.
+    pub fn post<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let matcher = HttpMatcher::method(MethodMatcher::POST);
+        self.match_route(path, matcher, service)
+    }
+
+    /// add a PUT route to the router.
+    pub fn put<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let matcher = HttpMatcher::method(MethodMatcher::PUT);
+        self.match_route(path, matcher, service)
+    }
+
+    /// add a DELETE route to the router.
+    pub fn delete<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let matcher = HttpMatcher::method(MethodMatcher::DELETE);
+        self.match_route(path, matcher, service)
+    }
+
+    /// add a PATCH route to the router.
+    pub fn patch<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let matcher = HttpMatcher::method(MethodMatcher::PATCH);
+        self.match_route(path, matcher, service)
+    }
+
+    /// add a HEAD route to the router.
+    pub fn head<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let matcher = HttpMatcher::method(MethodMatcher::HEAD);
+        self.match_route(path, matcher, service)
+    }
+
+    /// add a OPTIONS route to the router.
+    pub fn options<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let matcher = HttpMatcher::method(MethodMatcher::OPTIONS);
+        self.match_route(path, matcher, service)
+    }
+
+    /// add a TRACE route to the router.
+    pub fn trace<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let matcher = HttpMatcher::method(MethodMatcher::TRACE);
+        self.match_route(path, matcher, service)
+    }
+
+    /// add a CONNECT route to the router.
+    pub fn connect<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let matcher = HttpMatcher::method(MethodMatcher::CONNECT);
+        self.match_route(path, matcher, service)
+    }
+
+    /// register a nested router under a prefix.
+    ///
+    /// The prefix is used to match the request path and strip it from the request URI.
+    pub fn sub<I, T>(self, prefix: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let path = format!("{}/{}", prefix.trim_end_matches(['/']), "{*nest}");
+        let matcher = HttpMatcher::custom(true);
+
+        let nested_router_service = NestedRouterService {
+            prefix: prefix.to_owned(),
+            nested: service.into_endpoint_service().boxed(),
```

Comment:
I just realized, sorry for not seeing that before, that because we allow _any_ service, the concept of `ParentUriParams` is not sound. I think we need to remove `ParentUriParams` again and instead you need to make sure that _anywhere_ (so also in existing code from before this PR) where you make+inject a `UriParams` you need to have the following kind of logic:

```
let params: UriParams = ctx.remove::<UriParams>() {
      Some(params) => params.extend(new_params),
      None => new_params.collect(),
}
ctx.insert(params);
```

hope that's clear. That also means you want to some kind of `extend` feature that works with an iterator, similar to how `fn extend` is implemented for stuff such as `Vec`.